app/services/discourse_subscriptions/campaign.rb

```diff
@@ -0,0 +1,175 @@
+# frozen_string_literal: true
+
+module DiscourseSubscriptions
+  class Campaign
+    include DiscourseSubscriptions::Stripe
+
+    def initialize
+      set_api_key # instantiates Stripe API
+    end
+
+    def refresh_data
+      product_ids = Product.all.pluck(:external_id)
+
+      # if a product id is set for the campaign, we only want to return those results.
+      # if it's blank, return them all.
+      campaign_product = SiteSetting.discourse_subscriptions_campaign_product
+      product_ids = product_ids.select { |id| id == campaign_product } if campaign_product.present?
+
+      amount = 0
+      subscriptions = get_subscription_data
+      subscriptions = filter_to_subscriptions_products(subscriptions, product_ids)
+
+      # get number of subscribers
+      SiteSetting.discourse_subscriptions_campaign_subscribers = subscriptions&.length || 0
+
+      # calculate amount raised
+      subscriptions&.map do |sub|
+        sub_amount = calculate_monthly_amount(sub)
+        amount += sub_amount
+      end
+
+      SiteSetting.discourse_subscriptions_campaign_amount_raised = amount
+
+      if SiteSetting.discourse_subscriptions_campaign_show_contributors == true
+        contributor_ids = campaign_product.present? ? Customer.where(product_id: campaign_product).last(5).pluck(:user_id) : Customer.last(5).pluck(:user_id)
+        usernames = contributor_ids.map { |id| ::User.find(id).username }
+        SiteSetting.discourse_subscriptions_campaign_contributors = usernames.join(",") || ""
```

Comment:
I am not a huge fan of storing a list of usernames in a site setting like this. We might want to consider querying these contributors via a JSON endpoint rather than using a site setting. What happens if a user changes their username? or is deleted?

Reply:
I had considered this. The job updates the field every 30 minutes, so a deleted user or changed username won't be a major issue. The reason I went with the site setting is I'd have to build an endpoint just for this info. 

However, in thinking about it, I hadn't previously considered the N+1 happening in the `campaign-banner.js` `init()` function where we look up each user individually from the username. I thought this might be faster if the User model was cached, but I didn't see any instances of us using the store for users -- only topics and the like. It's a larger scope of a change, but I'll transition this to a dedicated route.

Reply:
Thanks I think it will be better for it.